CONTRIBUTING.md

```diff
@@ -1,69 +1,138 @@
+`cargo make build`.
 
-[cargo-make]: https://github.com/sagiegurari/cargo-make "cargo-make"
+```shell
+git clone https://github.com/tui-rs-revival/ratatui.git
+cd ratatui
+cargo make build
+```
 
-`ratatui` is an ordinary Rust project where common tasks are managed with [cargo-make].
-It wraps common `cargo` commands with sane defaults depending on your platform of choice.
-Building the project should be as easy as running `cargo make build`.
+### Tests
 
-## :hammer_and_wrench: Pull requests
+The [test coverage](https://app.codecov.io/gh/tui-rs-revival/ratatui) of the crate is reasonably
+good, but this can always be improved. Focus on keeping the tests simple and obvious and write unit
+tests for all new or modified code. Beside the usual doc and unit tests, one of the most valuable
+test you can write for `ratatui` is a test against the `TestBackend`. It allows you to assert the
+content of the output buffer that would have been flushed to the terminal after a given draw call.
+See `widgets_block_renders` in [tests/widgets_block.rs](./tests/widget_block.rs) for an example.
 
-All contributions are obviously welcome.
-Please include as many details as possible in your PR description to help the reviewer (follow the provided template).
-Make sure to highlight changes which may need additional attention or you are uncertain about.
-Any idea with a large scale impact on the crate or its users should ideally be discussed in a "Feature Request" issue beforehand.
+Prefer unit tests (and doc tests) in the code file over integration tests.
 
-## Committing
+If an area that you're making a change in is not tested, write tests to characterize the existing
+behavior before changing it. This helps ensure that we don't introduce bugs to existing software
+using ratatui (and helps make it easy to migrate apps still using tui-rs).
 
-To avoid any issues that may arrise with the CI/CD by not following the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) syntax, we recommend to install [Commitizen](https://commitizen-tools.github.io/commitizen/).\
-By using this tool you automatically follow the configuration defined in [.cz.toml](.cz.toml).
+### Use of unsafe for optimization purposes
 
-Additionally, we're using [cargo-husky](https://github.com/rhysd/cargo-husky) to automatically load pre-push hook, which will run `cargo make ci` before each push. It will load the hook automatically when you run `cargo test`. If `cargo-make` is not installed, it will install it for you.\
-This will ensure that your code is formatted, compiles and passes all tests before you push. If you want to skip this check, you can use `git push --no-verify`.
+**Do not** use unsafe to achieve better performances. This is [subject to
+change](https://github.com/tui-rs-revival/tui-rs-revival/discussions/66) The only exception to this
+rule is if it's to fix **reproducible slowness.**
```

Comment:
Need more info on what reproducible slowness means.

Reply:
My guess is that we can just drop this and say something about discussing the issue instead.